### ISSUE TITLE
jsonpath interprets index expressions incorrectly

### ISSUE THREAD

**Issue by rogpeppe** [kubernetes/client-go#1018]
The `jsonpath` package interprets `$.['a.b']` the same as `$.a.b` which is incorrect. It should be the same as `$.a\.b`.

Here's some code to demonstrate the issue: https://play.golang.org/p/qIaggWZVHl5

Also demonstrated by the above example: it should be OK to use a double-quoted string as an index expression, but it gives a syntax error. See https://www.ietf.org/id/draft-ietf-jsonpath-base-01.html#section-3.5.4.

**Comment by rogpeppe** [comment]
Having looked in a bit more detail at the jsonpath implementation, it really needs a very significant refactoring, if not a rewrite.

Here's an example from the code:

```
func (p *Parser) parseArray(cur *ListNode) error {
Loop:
	for {
		switch p.next() {
		case eof, '\n':
			return fmt.Errorf("unterminated array")
		case ']':
			break Loop
		}
	}
	text := p.consumeText()
```

`parseArray` is called after finding a `[` character. This code is assuming that _any_ `]` character will terminate the index expression. but that's not true when you can have a quoted string in there which can legitimately contain `]` characters itself.

Other places I observed issues:

- within the `parseArray` function again, it does this:

```
	value := dictKeyRex.FindStringSubmatch(text)
	if value != nil {
		parser, err := parseAction("arraydict", fmt.Sprintf(".%s", value[1]))
		if err != nil {
			return err
		}
```

`dictKeyRex` is the regular expression `^'([^']*)'$`. That is, if the regular expression matches, it is assuming a dict key index (a string). However it won't match any string that contains an escaped quote. Neither does it work with double-quoted strings.

Also, the above logic takes place _after_ splitting the index text on `,`, so a dict key index containing a comma won't work correctly either.

- inside `parseQuote`, it does this:
```
	for {
		switch p.next() {
		case eof, '\n':
			return fmt.Errorf("unterminated quoted string")
		case end:
			//if it's not escape break the Loop
			if p.input[p.pos-2] != '\\' {
				break Loop
			}
		}
	}
```

`end` is either `'` or `"`. That is, the first time it sees a quote character, it looks at the character before it, and if it's a backslach (`\`), it decides that the quote is escaped and continues. However there are times when a quote can be legitimately preceded by an escape character without being escaped, for example `'\\'`.

I'd suggest refactoring the code so it uses a conventional lexer-based approach with a recursive-descent parser layered on top of that.

**Comment by emmjohnson** [comment]
Another example of this, if the key of a map contains a space, it is unable to interpret the value. I've created a test that highlights the issue here - https://github.com/kubernetes/client-go/compare/master...emmjohnson:map-spaces?expand=1

**Comment by sthielo** [comment]
I also stumbled over this issue. As stated above there are 2 problems
* Any quoted string-literal used as name-selector is processed unquoted. So any special character (e.g. space, comma, brackets, ...) within a name-selector will cause a problem and is NOT behaving properly.
* Assumes only single quotes being used for string-literals used as name-selectors. 

This makes this library, which seems to be broadly used very unreliable for general json parsing. The following littel patch adds a simple test to show some of these problems.

```
Subject: [PATCH] adds a test for nasty characters
---
Index: util/jsonpath/jsonpath_test.go
===================================================================
diff --git a/util/jsonpath/jsonpath_test.go b/util/jsonpath/jsonpath_test.go
--- a/util/jsonpath/jsonpath_test.go	(revision 1517ffb8d37c99e6a3a2842bcdee0aa271f0332b)
+++ b/util/jsonpath/jsonpath_test.go	(date 1679843509883)
@@ -947,3 +947,70 @@
 		t,
 	)
 }
+
+func TestNastyChars(t *testing.T) {
+	var input = []byte(
+		`{
+			"foo": "bar",
+			"has space": "expected value",
+			"nested": {
+				"nested name": "nested value"
+			},
+			"has,comma": "expected comma value"
+		}`)
+
+	var data interface{}
+	err := json.Unmarshal(input, &data)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testJSONPath(
+		[]jsonpathTest{
+			{
+				"test top level property name containing space starting at root",
+				`{$['has space']}`,
+				data,
+				"expected value",
+				false,
+			},
+			{
+				"test top level property name containing space starting at current",
+				`{@['has space']}`,
+				data,
+				"expected value",
+				false,
+			},
+			{
+				"test child property name containing space in bracket notation",
+				`{$..['nested name']}`,
+				data,
+				"nested value",
+				false,
+			},
+			{
+				"test child property name containing space dot notation",
+				`{$.nested.'nested name'}`,
+				data,
+				"nested value",
+				false,
+			},
+			{
+				"test property name containing comma",
+				`{.'has,comma'}`,
+				data,
+				"expected comma value",
+				false,
+			},
+			{
+				"test double quoted name literal",
+				`{."foo"}`,
+				data,
+				"bar",
+				false,
+			},
+		},
+		false,
+		t,
+	)
+}

```

**Comment by sthielo** [comment]
I offer a draft of a re-implementation. I started with trying to do minor fixes and ended up with a complete re-implementation:
[https://github.com/sthielo/client-go/tree/jsonpath-issue-1018/util/jsonpath](https://github.com/sthielo/client-go/tree/jsonpath-issue-1018/util/jsonpath) (find TODOs in ./doc.go - e.g. nicer error messages, some more JSONPath filter functions, ...)

BUT ...
... it will break some - hopefully rare - cases (see TODOs in test files), as [JSONPath spec ](https://www.ietf.org/archive/id/draft-ietf-jsonpath-base-12.html) contradicts some of the examples. Also I tried to be resilient for backward compatibility not all can be made available without major issues.

I would be happy to have/discuss some reviews and votes from the community whether this would be welcome as a pull request.
